protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArrayBuilder.java

```diff
@@ -16,74 +16,67 @@
-        new ProtoArrayBuilder()
-            .justifiedCheckpoint(anchor.getCheckpoint())
-            .finalizedCheckpoint(anchor.getCheckpoint())
-            .build();
-    protoArray.onBlock(
-        anchor.getBlockSlot(),
-        anchor.getRoot(),
-        anchor.getParentRoot(),
-        anchor.getStateRoot(),
-        anchor.getEpoch(),
-        anchor.getEpoch());
-    return protoArray;
+  public ProtoArrayBuilder pruneThreshold(final int pruneThreshold) {
+    this.pruneThreshold = pruneThreshold;
+    return this;
   }
 
-  public ProtoArrayBuilder anchor(final Optional<Checkpoint> anchor) {
-    this.anchor = anchor;
+  public ProtoArrayBuilder initialEpoch(final UInt64 initialEpoch) {
+    this.initialEpoch = initialEpoch;
     return this;
   }
 
   public ProtoArrayBuilder justifiedCheckpoint(final Checkpoint justifiedCheckpoint) {
-    this.justifiedCheckpoint = justifiedCheckpoint;
+    checkNotNull(justifiedCheckpoint, "Justified checkpoint must be supplied");
+    this.justifiedEpoch(justifiedCheckpoint.getEpoch());
     return this;
   }
 
   public ProtoArrayBuilder finalizedCheckpoint(final Checkpoint finalizedCheckpoint) {
-    this.finalizedCheckpoint = finalizedCheckpoint;
+    checkNotNull(finalizedCheckpoint, "finalized checkpoint must be supplied");
+    this.finalizedEpoch(finalizedCheckpoint.getEpoch());
     return this;
   }
 
-  public ProtoArrayBuilder protoArraySnapshot(
-      final Optional<ProtoArraySnapshot> protoArraySnapshot) {
-    this.protoArraySnapshot = protoArraySnapshot;
+  public ProtoArrayBuilder justifiedEpoch(final UInt64 justifiedEpoch) {
+    this.justifiedEpoch = justifiedEpoch;
     return this;
   }
 
-  public ProtoArray build() {
-    checkNotNull(justifiedCheckpoint, "Justified checkpoint must be supplied");
-    checkNotNull(finalizedCheckpoint, "Finalized checkpoint must be supplied");
+  public ProtoArrayBuilder finalizedEpoch(final UInt64 finalizedEpoch) {
+    this.finalizedEpoch = finalizedEpoch;
+    return this;
+  }
 
-    // If no anchor is explicitly set, default to zero (genesis epoch)
-    final UInt64 anchorEpoch =
-        anchor.map(Checkpoint::getEpoch).orElse(UInt64.valueOf(Constants.GENESIS_EPOCH));
+  public ProtoArrayBuilder initialCheckpoint(final Optional<Checkpoint> initialCheckpoint) {
+    initialCheckpoint.ifPresentOrElse(
+        checkpoint -> initialEpoch(checkpoint.getEpoch()),
+        () -> initialEpoch(SpecConstants.GENESIS_EPOCH));
+    return this;
+  }
 
-    return protoArraySnapshot
-        .map(ProtoArraySnapshot::toProtoArray)
-        .orElse(
-            new ProtoArray(
-                Constants.PROTOARRAY_FORKCHOICE_PRUNE_THRESHOLD,
-                justifiedCheckpoint.getEpoch(),
-                finalizedCheckpoint.getEpoch(),
-                anchorEpoch,
-                new ArrayList<>(),
-                new HashMap<>()));
+  ProtoArrayBuilder nodes(final List<ProtoNode> nodes) {
+    this.nodes = nodes;
+    return this;
```

Comment:
It looks like we only ever pass in an empty list here and then only for two tests that want access to the list of nodes.  Probably should remove this method and not pass a list of nodes to the `ProtoArray` constructor at all.  The tests can just use `getNodes()` to get the list for their assertions.  That would resolve my question about where the indices get created for initially present nodes.